addon/PageScraper.js

```diff
@@ -0,0 +1,96 @@
+/* globals Services, Task */
+"use strict";
+
+const {MetadataParser} = require("addon/MetadataParser");
+const {Cu} = require("chrome");
+const options = require("@loader/options");
+
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource://gre/modules/Task.jsm");
+Cu.importGlobalProperties(["URL"]);
+
+const DEFAULT_OPTIONS = {
+  framescriptPath: new URL("data/page-scraper-content-script.js", options.prefixURI),
+  blacklist: ["about:", "localhost:", "resource://"]
+};
+
+function PageScraper(previewProvider, options = {}) {
+  this.options = Object.assign({}, DEFAULT_OPTIONS, options);
+  this._previewProvider = previewProvider;
+  this._metadataParser = new MetadataParser();
+}
+
+/**
+ * Receives raw HTML from a page and sends it to a service to be parsed and the
+ * metadata extracted, and then stores that in the metadata DB. It takes an instance
+ * of PreviewProvider, since all the link processing takes places in PreviewProvider
+ */
+
+PageScraper.prototype = {
+  /**
+   * Check if the link is already in the metadata database. If not, parse the
+   * HTML and insert it in the metadata database
+   */
+  _asyncParseAndSave: Task.async(function*(rawHTML, url) {
+    let link = yield this._previewProvider.asyncLinkExist(url);
+    if (!link) {
+      let metadata;
+      try {
+        metadata = yield this._metadataParser.parseHTMLText(rawHTML, url);
+      } catch (e) {
+        Cu.reportError(`MetadataParser failed to parse ${url}`);
```

Comment:
Make sure that logging actual URLs is only done locally (ie the browser console).